h2o-algos/src/main/java/hex/tree/Score.java

```diff
@@ -17,19 +19,33 @@
   final boolean _oob;           // Computed on OOB
   final Key<Vec> _kresp;        // Response vector key (might be either train or validation)
   final ModelCategory _mcat;    // Model category (Binomial, Regression, etc)
-  ModelMetrics.MetricBuilder _mb;
-//  GainsLift.GainsLiftBuilder _gainsLiftBuilder;
+  ModelMetricsSupervised.MetricBuilderSupervised _mb;
   final boolean _computeGainsLift;
+  final ScoreIncInfo _sii;      // Incremental scoring (on a validation dataset), null indicates full scoring
+  final Frame _preds;           // Prediction cache (typically not too many Vecs => it is not too costly embed the object in MRTask)
 
   /** Compute ModelMetrics on the testing dataset.
    *  It expect already adapted validation dataset which is adapted to a model
    *  and contains a response which is adapted to confusion matrix domain.
    */
-  public Score(SharedTree bldr, boolean is_train, boolean oob, Key<Vec> kresp, ModelCategory mcat, boolean computeGainsLift) { _bldr = bldr; _is_train = is_train; _oob = oob; _kresp = kresp; _mcat = mcat; _computeGainsLift = computeGainsLift; }
+  public Score(SharedTree bldr, boolean is_train, boolean oob, Vec kresp, ModelCategory mcat, boolean computeGainsLift, Frame preds) {
+    this(bldr, is_train, null, oob, kresp, mcat, computeGainsLift, preds);
+  }
+
+  public Score(SharedTree bldr, ScoreIncInfo sii, boolean oob, Vec kresp, ModelCategory mcat, boolean computeGainsLift, Frame preds) {
+    this(bldr, false, sii, oob, kresp, mcat, computeGainsLift, preds);
+  }
+
+  private Score(SharedTree bldr, boolean is_train, ScoreIncInfo sii, boolean oob, Vec kresp, ModelCategory mcat, boolean computeGainsLift, Frame preds) {
+    _bldr = bldr; _is_train = is_train; _sii = sii; _oob = oob; _kresp = kresp._key; _mcat = mcat; _computeGainsLift = computeGainsLift;
+    _preds = computeGainsLift ? preds : null; // don't keep the prediction cache if we don't need to compute gainslift
+    assert (! _is_train) || (_sii == null);
+  }
 
-  @Override public void map( Chunk chks[] ) {
+  @Override public void map(Chunk allchks[]) {
+    final Chunk[] chks = getScoringChunks(allchks);
     Chunk ys = _bldr.chk_resp(chks);  // Response
-    Model m = _bldr._model;
+    SharedTreeModel m = _bldr._model;
```

Comment:
Side note (unrelated from the change): having model serialized is really bad, i still do not know why we need full model, not only selected info.

Reply:
This is a good point. We don't need the whole model object